### PR TITLE
Added flyteidl documentation

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
Signed-off-by: Prafulla Mahindrakar <prafulla.mahindrakar@gmail.com>

# TL;DR
Added documentation for flyteidl using protoc gen doc and removed the documentation using protodoc
https://github.com/pseudomuto/protoc-gen-doc

**The DOC are generated in the following steps.**
1] Make doc_gen_deps : This downloads all the dependencies required for the doc generation in tmp folder.(excluded from git)
2] generate_protos.sh : This generates the IDL for different languages using the LYFT protocgenerator image 
    And later also generates the IDL docs as Markdown files ( protoc-gen-doc currently doesn;t support rst)
3] The generated Markdown files are then used to generate html docs using sphinx.

Note : Have removed the existing way of generating the documentation and its corresponding files.

**Make file changes:**
Introduced a new doc_gen_deps step in the Makefile which is required to be run before generating the protos inside which the docs are generated.

This steps downloads dependencies which are required by protoc gen doc for the protos which have external library dependencies which includes grpc-gateway, googleapis, k8.io/api and apimachinery, protocolbuffers.

The downloaded deps are mounted as volumes on the protoc-gen-doc docker container so that the definitions are available to complete the MD file generation of all the protos. (without which protoc-gen-doc throws a fatal error and aborts without generating the MD file)

This is not required for IDL generation as its using the LYFT protocgenerator image which includes these dependencies through --with-gateway option. Once we move away from the LYFT image , these steps can be copied over into a common image which can do both IDL generation and DOC generation

**Doc changes are available here**
https://docs.flyte.org/projects/flyteidl/en/docs-revamp/protos/index.html#

**And can be compare with this** 
https://docs.flyte.org/projects/flyteidl/en/latest/gen/pb-protodoc/flyteidl/index.html

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/<number>

## Follow-up issue
Cross linking of the idl refences.
eg : href #flyteidl.core.Identifier  is generated by the docs in the core folder.
Due to the docs segregation , the links to flyteidl.core.Identifier are not resolved correctly as they try to find the definition in the same doc. A post processing step is required to fix such links.

### CODE REVIEW

**Comment by kumare3** [comment]
@cosmicBboy @samhita-alla, this might affect intersphinx references

**Comment by cosmicBboy** [comment]
> @cosmicBboy @samhita-alla, this might affect intersphinx references

yeah, looks like there are a couple of places that reference flyteidl in `flyte` and `flytesnacks`... not sure which are related to changes in this PR tho:

#### In `flyte`
```
flyte/rsts/community/contribute.rst
44:#. `flyteidl <https://github.com/lyft/flyteidl>`_
47:    | Setup: Refer to the `README <https://github.com/flyteorg/flyteidl#generate-code-from-protobuf>`_

flyte/rsts/community/roadmap.rst
19:Flyte consists of many components and services. In true Agile fashion each service is independently iterated and co-ordiated by maintaing backwards compatible contracts using protobuf defined in `idl <https://flyte.readthedocs.io/projects/flyteidl/en/latest/>`__. Thus components like flytekit, flytepropeller, datacatalog are independently versioned.

flyte/rsts/concepts/admin.rst
21:service `definition <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/service/admin.proto>`__. The
62:`here <https://github.com/flyteorg/flyteidl/tree/master/protos/flyteidl/admin>`__.

flyte/rsts/concepts/admin_service.rst
9:The `admin service definition <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/service/admin.proto>`__ defines REST operations for the entities
29:The above are designated by an `identifier <https://github.com/flyteorg/flyteidl/blob/793b09d190148236f41ad8160b5cec9a3325c16f/protos/flyteidl/core/identifier.proto#L16>`_
33:One caveat is that launch plan state can toggle between `ACTIVE or INACTIVE <https://github.com/flyteorg/flyteidl/blob/6b8e34af67cbd06d2f20a6f7462e12c7f0d34f3b/protos/flyteidl/admin/launch_plan.proto#L33>`_.
40:A `named entity <https://github.com/flyteorg/flyteidl/blob/c12816cf3f0bde54a67fa77e71c0158a64151ea3/protos/flyteidl/admin/common.proto#L45>`__ includes metadata for one of the above entities
42:`id <https://github.com/flyteorg/flyteidl/blob/c12816cf3f0bde54a67fa77e71c0158a64151ea3/protos/flyteidl/admin/common.proto#L12>`__ which is composed of project, domain and name.
75:These `events <https://github.com/flyteorg/flyteidl/blob/793b09d190148236f41ad8160b5cec9a3325c16f/protos/flyteidl/event/event.proto>`_ include
177:     - States are defined in `launch_plan.proto <https://github.com/flyteorg/flyteidl/blob/2c17791170ece1cced3e96daa08ea2692efe3d07/protos/flyteidl/admin/launch_plan.proto#L23>`_.
181:     - States are defined in `common.proto <https://github.com/flyteorg/flyteidl/blob/c12816cf3f0bde54a67fa77e71c0158a64151ea3/protos/flyteidl/admin/common.proto#L25>`_.
190:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L11,L21>`__.
195:     - Modes are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/182eeb9e1d0f2369e479a981f30cb51c2d7a0672/protos/flyteidl/admin/execution.proto#L96>`__.
203:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L26,L36>`__.
215:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L42,L49>`__.
271:     - States are defined in `launch_plan.proto <https://github.com/flyteorg/flyteidl/blob/2c17791170ece1cced3e96daa08ea2692efe3d07/protos/flyteidl/admin/launch_plan.proto#L23>`_.
282:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L11,L21>`__.
287:     - Modes are defined `execution.proto <https://github.com/flyteorg/flyteidl/blob/182eeb9e1d0f2369e479a981f30cb51c2d7a0672/protos/flyteidl/admin/execution.proto#L96>`__.
293:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L26,L36>`__.
302:     - Phases are defined in `execution.proto <https://github.com/flyteorg/flyteidl/blob/223537e15e05bc6925403a8c11c5a09d91008a80/protos/flyteidl/core/execution.proto#L42,L49>`__.

flyte/rsts/concepts/architecture.rst
7:This document aims to demystify how Flyte's major components ``FlyteIDL``, ``FlyteKit``, ``FlyteCLI``, ``FlyteConsole``, ``FlyteAdmin``, ``FlytePropeller``, and ``FlytePlugins`` fit together at a high level.
9:FlyteIDL
14:The Flyte IDL (Interface Definition Language) is where shared Flyte entities are defined. This IDL also defines the RPC service definition for the `core Flyte API <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/service/admin.proto>`_.
16:FlyteIDL uses the `protobuf <https://developers.google.com/protocol-buffers/>`_ schema to describe entities. Clients are generated for Python, Golang, and JavaScript and imported by Flyte components.
59:The Control Plane supports the core REST/gRPC API defined in FlyteIDL. User Plane tools like FlyteConsole and FlyteKit contact the control plane on behalf of users to store and retrieve information.

flyte/rsts/concepts/customizable_resources.rst
7:For background on customizable resources, see :ref:`howto-managing-customizable-resources`. As a quick refresher, custom resources allow you to manage configurations for specific combinations of user projects, domains and workflows that override default values. Examples of such resources include execution clusters, task resource defaults, and `more <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/matchable_resource.proto>`__.
19:You'll want to introduce a new `matchable attribute <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/matchable_resource.proto>`__, including a unique enum value and proto message definition. For example
40:See the changes in this `file <https://github.com/flyteorg/flyteidl/commit/b1767697705621a3fddcb332617a5304beba5bec#diff-d3c1945436aba8f7a76755d75d18e671>`__ for an example of what is required.

flyte/rsts/concepts/registration.rst
14: * The user then uses flytekit's register cli to compile the tasks into their serialized representation as described in :std:ref:`Flyte Specification language <flyteidl:flyteidltoc>`. During this time the task representation is also bound to a container that contains the code for the task. This associated entity is registered with FlyteAdmin using the registerTask api.
15: * The user then uses flytekit's register cli to compile the workflow into their serialized representation as described in :std:ref:`Flyte Specification language <flyteidl:flyteidltoc>`. The referenced tasks are replaced by their Flyte Admin registered Identifier, obtained in the previous step. This associated entity is registered with FlyteAdmin using the registerWorkflow api.

flyte/rsts/concepts/state_machine.rst
29:Every transition between states is recorded in Flyteadmin using :std:ref:`gen/pb-protodoc/flyteidl/event/event.proto:flyteidl.event.workflowexecutionevent`
31:The phases in the above state diagram are captured in the Admin database as specified here :std:ref:`api_enum_flyteidl.core.workflowexecution.phase` and are sent as part of the Execution Event.
50:#. :std:ref:`gen/pb-protodoc/flyteidl/core/workflow.proto:flyteidl.core.tasknode`
51:#. :std:ref:`gen/pb-protodoc/flyteidl/core/workflow.proto:flyteidl.core.branchnode`
52:#. :std:ref:`gen/pb-protodoc/flyteidl/core/workflow.proto:flyteidl.core.workflownode`
56:Every transition between states is recorded in Flyteadmin using :std:ref:`gen/pb-protodoc/flyteidl/event/event.proto:flyteidl.event.nodeexecutionevent`
58:Every NodeExecutionEvent can have one of the :std:ref:`api_enum_flyteidl.core.nodeexecution.phase`
72:Every transition between states is recorded in Flyteadmin using :std:ref:`gen/pb-protodoc/flyteidl/event/event.proto:flyteidl.event.taskexecutionevent`
74:Every TaskExecutionEvent can have one of the :std:ref:`api_enum_flyteidl.core.taskexecution.phase`

flyte/rsts/concepts/tasks.rst
40:   signature in programming languages). A task interface defines the input and output variables - :std:ref:`flyteidl:api_field_flyteidl.core.variablemap.variables`
41:   as well as their types :std:ref:`flyteidl:api_msg_flyteidl.core.literaltype`.

flyte/rsts/howto/managing_customizable_resources.rst
28:`matchable attributes <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/matchable_resource.proto>`_
175:Use the `get <https://github.com/flyteorg/flyteidl/blob/ba13965bcfbf7e7bfce40664800aaf1f2a1088a1/protos/flyteidl/service/admin.proto#L395>`__ endpoint
182:The resource type enum (int) is defined in the `proto <https://github.com/flyteorg/flyteidl/blob/ba13965bcfbf7e7bfce40664800aaf1f2a1088a1/protos/flyteidl/admin/matchable_resource.proto#L8,L20>`__.

flyte/rsts/reference/index.rst
15:   FlyteIDL <https://flyteidl.readthedocs.io>
```

### In `flytesnacks`

```
flytesnacks/cookbook/core/extend_flyte/backend_plugins.rst
26:Usually Flyte extensions need information that is not covered by a :std:ref:`Flyte TaskTemplate <api_msg_flyteidl.core.tasktemplate>`. The TaskTemplate consists of a
27:the interface, task_type identifier, some metadata and other fields. An important field to note here is - :std:ref:`api_field_flyteidl.core.tasktemplate.custom`. The custom field is essentially an unstructured JSON.
28:This makes it possible to extend a task-template beyond the default supported targets -- :std:ref:`api_field_flyteidl.core.tasktemplate.container` (WIP, sql etc).
33:For Spark we decided to use Protobuf to specify the plugin as can be seen `here <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/plugins/spark.proto>`__. Note it is not necessary to have the Plugin structure specified in FlyteIDL, we do it for simplicity, ease of maintenance alongwith the core platform and because of existing tooling to generate code for protobuf.

flytesnacks/cookbook/docs/auto/core/control_flow/index.rst
61:    <div class="sphx-glr-thumbcontainer" tooltip="FlyteIdl (the fundamental building block of the Flyte Language) allows various programming lang...">

flytesnacks/cookbook/docs/auto/core/control_flow/run_merge_sort.rst
26:FlyteIdl (the fundamental building block of the Flyte Language) allows various programming language features;

flytesnacks/cookbook/docs/auto/core/flyte_basics/files.rst
25:In the IDL, they are known as `Blob <https://github.com/lyft/flyteidl/blob/cee566b2e6e109120f1bb34c980b1cfaf006a473/protos/flyteidl/core/literals.proto#L33>`__ literals
26:and are backed by the `blob type <https://github.com/lyft/flyteidl/blob/cee566b2e6e109120f1bb34c980b1cfaf006a473/protos/flyteidl/core/types.proto#L47>`__

flytesnacks/cookbook/docs/auto/core/flyte_basics/folders.rst
24:Please also see the entry on files. After files, folders are the other fundamental operating system primitive users might find themselves working with. The Flyte IDL's support of folders take the form of `multi-part blobs <https://github.com/lyft/flyteidl/blob/cee566b2e6e109120f1bb34c980b1cfaf006a473/protos/flyteidl/core/types.proto#L50>`__.

flytesnacks/cookbook/docs/extend_flyte_backend_plugins.rst
26:Usually Flyte extensions need information that is not covered by a :std:ref:`Flyte TaskTemplate <api_msg_flyteidl.core.tasktemplate>`. The TaskTemplate consists of a
27:the interface, task_type identifier, some metadata and other fields. An important field to note here is - :std:ref:`api_field_flyteidl.core.tasktemplate.custom`. The custom field is essentially an unstructured JSON.
28:This makes it possible to extend a task-template beyond the default supported targets -- :std:ref:`api_field_flyteidl.core.tasktemplate.container` (WIP, sql etc).
33:For Spark we decided to use Protobuf to specify the plugin as can be seen `here <https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/plugins/spark.proto>`__. Note it is not necessary to have the Plugin structure specified in FlyteIDL, we do it for simplicity, ease of maintenance alongwith the core platform and because of existing tooling to generate code for protobuf.
```

**Comment by cosmicBboy** [comment]
I'd love to get the error/warning count to our docs-building processes so we can catch stale intersphinx refs

**Comment by samhita-alla** [comment]
> I'd love to get the error/warning count to our docs-building processes so we can catch stale intersphinx refs

PR: https://github.com/flyteorg/flytesnacks/pull/222

**Comment by pmahindrakar-oss** [comment]
PR's for intersphinx link changes in flytesnacks and flyte

https://github.com/flyteorg/flytesnacks/pull/223
https://github.com/flyteorg/flyte/pull/1042

The definitions are being pulled from https://docs.flyte.org/projects/flyteidl/en/latest
So the links would be effective after this change reaches production

**Review by kumare3 — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -38,6 +38,23 @@ build_python:
 install-piptools:
 	pip install -U pip-tools
 
+.PHONY: doc_gen_deps # these dependencies are required by protoc gen doc for the protos which have external library dependencies. 
+# which includes grpc-gateway, googleapis, k8.io/api and apimachinery, protocolbuffers 
+doc_gen_deps:
+	rm -rf tmp/doc_gen_deps tmp/protocolbuffers tmp/grpc-gateway tmp/k8s.io
```
Comment (kumare3) [inline Makefile:44]: can we move this to a script?

Comment (pmahindrakar-oss) [reply Makefile:44]: Yes taking this followup where i am fixing the link  which are refrenced from other modules

**Review by samhita-alla — APPROVED** [review]
